packages/api-client-core/src/support.ts

```diff
@@ -543,3 +545,87 @@ const checkEquality = (a: any, b: any, refs: any[]): boolean => {
 };
 
 export const isEqual = (a: any, b: any) => checkEquality(a, b, []);
+
+/**
+ * Processes the flexible, convenient JS-land inputs for an action to the fully qualified GraphQL API inputs
+ * @internal
+ **/
+export const disambiguateActionVariables = (action: AnyActionFunction, variables: Record<string, any> | undefined) => {
+  variables ??= {};
+  if (!("hasAmbiguousIdentifier" in action) && !("acceptsModelInput" in action)) return variables;
+
+  if (action.hasAmbiguousIdentifier) {
+    if (
+      Object.keys(variables).some((key) => key !== "id" && !action.paramOnlyVariables?.includes(key) && key !== action.modelApiIdentifier)
+    ) {
+      throw Error(`Invalid arguments found in variables. Did you mean to use ({ ${action.modelApiIdentifier}: { ... } })?`);
+    }
+  }
+
+  let newVariables: Record<string, any>;
+
+  if (action.acceptsModelInput ?? action.hasCreateOrUpdateEffect) {
+    if (
+      action.modelApiIdentifier in variables &&
+      typeof variables[action.modelApiIdentifier] === "object" &&
+      variables[action.modelApiIdentifier] != null
+    ) {
+      newVariables = variables;
+    } else {
+      newVariables = {
+        [action.modelApiIdentifier]: {},
+      };
+      for (const [key, value] of Object.entries(variables)) {
+        if (action.paramOnlyVariables?.includes(key)) {
+          newVariables[key] = value;
+        } else {
+          if (key == "id") {
+            newVariables.id = value;
+          } else {
+            newVariables[action.modelApiIdentifier][key] = value;
+          }
+        }
+      }
+    }
+  } else {
+    newVariables = variables;
+  }
+
+  return newVariables;
+};
```

Comment:
This helper used to exist in the react package, but I needed it in api client core now, so I moved it here.